src/datastore/postgres-store.ts

```diff
@@ -881,42 +881,42 @@ export class PgDataStore
   }
 
   async storeRawEventRequest(eventPath: string, payload: string): Promise<void> {
-    await this.query(async client => {
-      const insertResult = await client.query<{ id: string }>(
-        `
-        INSERT INTO event_observer_requests(
+    // To avoid depending on the DB more than once and to allow the query transaction to settle,
+    // we'll take the complete insert result and move that to the output TSV file instead of taking
+    // only the `id` and performing a `COPY` of that row later.
+    const insertResult = await this.queryTx(async client => {
+      return await client.query<{
+        id: string;
+        receive_timestamp: string;
+        event_path: string;
+        jsonb: string;
+      }>(
+        `INSERT INTO event_observer_requests(
           event_path, payload
         ) values($1, $2)
-        RETURNING id
-        `,
+        RETURNING *`,
         [eventPath, payload]
       );
-      if (insertResult.rowCount !== 1) {
-        throw new Error(
-          `Unexpected row count ${insertResult.rowCount} when storing event_observer_requests entry`
-        );
-      }
-      const exportEventsFile = process.env['STACKS_EXPORT_EVENTS_FILE'];
-      if (exportEventsFile) {
-        const writeStream = fs.createWriteStream(exportEventsFile, {
-          flags: 'a', // append or create if not exists
-        });
-        try {
-          const queryStream = client.query(
-            pgCopyStreams.to(
-              `COPY (SELECT * FROM event_observer_requests WHERE id = ${insertResult.rows[0].id}) TO STDOUT ENCODING 'UTF8'`
-            )
-          );
-          await pipelineAsync(queryStream, writeStream);
-        } finally {
-          writeStream.close();
-        }
-      }
     });
+    if (insertResult.rowCount !== 1) {
+      throw new Error(
+        `Unexpected row count ${insertResult.rowCount} when storing event_observer_requests entry`
+      );
+    }
+    const exportEventsFile = process.env['STACKS_EXPORT_EVENTS_FILE'];
+    if (exportEventsFile) {
+      const result = insertResult.rows[0];
+      const tsvRow = [result.id, result.receive_timestamp, result.event_path, result.jsonb];
+      fs.appendFileSync(exportEventsFile, tsvRow.join('\t'));
```

Comment:
What do you think about this approach @zone117x ? Basically, I'm trying to finalize the query before writing synchronously into the TSV file.

My only worry here is that there could be a race condition for example if we get a barrage of events (like mempool txs), all the queries settle correctly, but then the `appendFileSync` calls fall out of order and the TSV file ends up with unordered rows. That is a possibility with the current code too, I believe, but it would only happen if we're running several processes per write-only pod.

Reply:
Cool this looks good to me, seems like best way to reduce odds of race conditions. Ultimately, it will still be possible for the process to crash or be forced killed between a sql commit and file write, but no way to really get around that.